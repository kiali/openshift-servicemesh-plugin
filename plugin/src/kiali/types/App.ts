import { Namespace } from './Namespace';
import { Runtime } from './Workload';
import { AppHealthResponse } from '../types/Health';

export interface AppId {
  cluster?: string;
  namespace: string;
  app: string;
}

export interface AppWorkload {
  workloadName: string;
  istioSidecar: boolean;
  istioAmbient: boolean;
  serviceAccountNames: string[];
  labels: { [key: string]: string };
}

export interface App {
  cluster?: string;
  namespace: Namespace;
  name: string;
  workloads: AppWorkload[];
  serviceNames: string[];
  runtimes: Runtime[];
  health: AppHealthResponse;
}
