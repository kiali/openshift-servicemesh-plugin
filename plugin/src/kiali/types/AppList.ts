import { Namespace } from './Namespace';
import { AppHealth } from './Health';
import { ObjectReference } from './IstioObjects';

export interface AppList {
  namespace: Namespace;
  applications: AppOverview[];
}

export interface AppOverview {
  name: string;
  cluster?: string;
  istioSidecar: boolean;
  istioAmbient: boolean;
  labels: { [key: string]: string };
  istioReferences: ObjectReference[];
  health: AppHealth;
}

export interface AppListItem extends AppOverview {
  namespace: string;
}
