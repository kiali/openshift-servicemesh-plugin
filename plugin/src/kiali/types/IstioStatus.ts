import { t } from 'utils/I18nUtils';

export enum Status {
  Healthy = 'Healthy',
  Unhealthy = 'Unhealthy',
  Unreachable = 'Unreachable',
  NotFound = 'NotFound',
  NotReady = 'NotReady'
}

export const statusMsg = {
  [Status.Healthy]: t('Healthy'),
  [Status.NotFound]: t('Not found'),
  [Status.NotReady]: t('Not ready'),
  [Status.Unhealthy]: t('Not healthy'),
  [Status.Unreachable]: t('Unreachable')
};

export interface ComponentStatus {
  is_core: boolean;
  name: string;
  status: Status;
}

export interface IstiodResourceThresholds {
  cpu: number;
  memory: number;
}
