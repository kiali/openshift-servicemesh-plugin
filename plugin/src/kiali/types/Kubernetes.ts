export interface TypeMeta {
  apiVersion: string;
  kind: string;
}
