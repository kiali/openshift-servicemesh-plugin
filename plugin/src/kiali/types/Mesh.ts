import { Controller, ElementModel, GraphElement } from '@patternfly/react-topology';
import { AppenderString } from './Common';

export interface MeshCluster {
  accessible: boolean;
  apiEndpoint: string;
  isKialiHome: boolean;
  kialiInstances: KialiInstance[];
  name: string;
  network: string;
  secretName: string;
}

export interface KialiInstance {
  namespace: string;
  operatorResource: string;
  serviceName: string;
  url: string;
  version: string;
}

export type MeshClusters = MeshCluster[];

// MESH GRAPH

export enum MeshInfraType {
  CLUSTER = 'cluster',
  GRAFANA = 'grafana',
  ISTIOD = 'istiod',
  KIALI = 'kiali',
  METRIC_STORE = 'metricStore',
  NAMESPACE = 'namespace',
  TRACE_STORE = 'traceStore'
}

export enum MeshNodeType {
  BOX = 'box',
  INFRA = 'infra'
}

// TODO
export type MeshNodeHealthData = string;

// Node data expected from server
export interface MeshNodeData {
  // required
  cluster: string;
  id: string;
  infraName: string;
  infraType: MeshInfraType;
  namespace: string;
  nodeType: MeshNodeType;

  // optional
  healthData?: MeshNodeHealthData;
  infraData?: MeshCluster; // add other type options as the case arises
  isAmbient?: boolean;
  isBox?: string;
  isInaccessible?: boolean;
  isMTLS?: boolean;
  isOutOfMesh?: boolean;
  labels?: { [key: string]: string };
  parent?: string;
}

// Edge data expected from server
export interface MeshEdgeData {
  id: string;
  source: string;
  target: string;
}

export interface MeshNodeWrapper {
  data: MeshNodeData;
}

export interface MeshEdgeWrapper {
  data: MeshEdgeData;
}

export interface MeshElements {
  edges?: MeshEdgeWrapper[];
  nodes?: MeshNodeWrapper[];
}

// TODO: unnecessary?
export interface MeshQuery {
  appenders?: AppenderString;
}

export interface MeshDefinition {
  elements: MeshElements;
  timestamp: number;
}

// Node data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedMeshNodeData extends MeshNodeData {
  healthStatus: string; // status name
}

// Edge data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedMeshEdgeData extends MeshEdgeData {
  // Default value -1
  isMTLS: number;

  // assigned when graph is updated, the edge health depends on the node health, traffic, and config
  healthStatus?: string; // status name
}

export interface DecoratedMeshNodeWrapper {
  data: DecoratedMeshNodeData;
}

export interface DecoratedMeshEdgeWrapper {
  data: DecoratedMeshEdgeData;
}

export interface DecoratedMeshElements {
  edges?: DecoratedMeshEdgeWrapper[];
  nodes?: DecoratedMeshNodeWrapper[];
}

export type MeshType = 'mesh' | 'node' | 'edge' | 'box';

export interface MeshTarget {
  elem: Controller | GraphElement<ElementModel, any> | undefined;
  type: MeshType; // the element type
}

export const MeshAttr = {
  // shared attrs
  id: 'id',

  // edge attrs
  healthStatus: 'healthStatus',
  isMTLS: 'isMTLS',

  // node attrs
  cluster: 'cluster',
  health: 'health',
  infraData: 'infraData',
  infraName: 'infraName',
  infraType: 'infraType',
  isInaccessible: 'isInaccessible',
  isOutOfMesh: 'isOutOfMesh',
  namespace: 'namespace',
  nodeType: 'nodeType'
};
