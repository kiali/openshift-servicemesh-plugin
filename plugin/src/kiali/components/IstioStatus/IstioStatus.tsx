import * as React from 'react';
import { SVGIconProps } from '@patternfly/react-icons/dist/esm/createIcon';
import * as API from '../../services/Api';
import * as AlertUtils from '../../utils/AlertUtils';
import { TimeInMilliseconds } from '../../types/Common';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { MessageType } from '../../types/MessageCenter';
import { Namespace } from '../../types/Namespace';
import { KialiAppState } from '../../store/Store';
import { istioStatusSelector, namespaceItemsSelector } from '../../store/Selectors';
import { bindActionCreators } from 'redux';
import { IstioStatusActions } from '../../actions/IstioStatusActions';
import { connect } from 'react-redux';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import { IstioStatusList } from './IstioStatusList';
import { PFColors } from '../Pf/PfColors';
import './IstioStatus.css';
import { ResourcesFullIcon } from '@patternfly/react-icons';
import { KialiDispatch } from 'types/Redux';
import { NamespaceThunkActions } from '../../actions/NamespaceThunkActions';
import { connectRefresh } from '../Refresh/connectRefresh';

type ReduxProps = {
  setIstioStatus: (istioStatus: ComponentStatus[]) => void;
  refreshNamespaces: () => void;
  namespaces: Namespace[] | undefined;
  status: ComponentStatus[];
};

type StatusIcons = {
  ErrorIcon?: React.ComponentClass<SVGIconProps>;
  WarningIcon?: React.ComponentClass<SVGIconProps>;
  InfoIcon?: React.ComponentClass<SVGIconProps>;
  HealthyIcon?: React.ComponentClass<SVGIconProps>;
};

type Props = ReduxProps & {
  lastRefreshAt: TimeInMilliseconds;
  icons?: StatusIcons;
  cluster?: string;
};

const ValidToColor = {
  'true-true-true': PFColors.Danger,
  'true-true-false': PFColors.Danger,
  'true-false-true': PFColors.Danger,
  'true-false-false': PFColors.Danger,
  'false-true-true': PFColors.Warning,
  'false-true-false': PFColors.Warning,
  'false-false-true': PFColors.Info,
  'false-false-false': PFColors.Success
};

const defaultIcons = {
  ErrorIcon: ResourcesFullIcon,
  WarningIcon: ResourcesFullIcon,
  InfoIcon: ResourcesFullIcon,
  HealthyIcon: ResourcesFullIcon
};

export class IstioStatusComponent extends React.Component<Props> {
  componentDidMount() {
    this.props.refreshNamespaces();
    this.fetchStatus();
  }

  componentDidUpdate(prevProps: Readonly<Props>): void {
    if (this.props.lastRefreshAt !== prevProps.lastRefreshAt) {
      this.fetchStatus();
    }
  }

  fetchStatus = () => {
    API.getIstioStatus(this.props.cluster)
      .then(response => {
        return this.props.setIstioStatus(response.data);
      })
      .catch(error => {
        // User without namespaces can't have access to mTLS information. Reduce severity to info.
        const informative = this.props.namespaces && this.props.namespaces.length < 1;
        if (informative) {
          AlertUtils.addError('Istio deployment status disabled.', error, 'default', MessageType.INFO);
        } else {
          AlertUtils.addError('Error fetching Istio deployment status.', error, 'default', MessageType.ERROR);
        }
      });
  };

  tooltipContent = () => {
    return <IstioStatusList status={this.props.status} />;
  };

  tooltipColor = () => {
    let coreUnhealthy: boolean = false;
    let addonUnhealthy: boolean = false;
    let notReady: boolean = false;

    Object.keys(this.props.status || {}).forEach((compKey: string) => {
      const { status, is_core } = this.props.status[compKey];
      const isNotReady: boolean = status === Status.NotReady;
      const isUnhealthy: boolean = status !== Status.Healthy && !isNotReady;

      if (is_core) {
        coreUnhealthy = coreUnhealthy || isUnhealthy;
      } else {
        addonUnhealthy = addonUnhealthy || isUnhealthy;
      }

      notReady = notReady || isNotReady;
    });

    return ValidToColor[`${coreUnhealthy}-${addonUnhealthy}-${notReady}`];
  };

  healthyComponents = () => {
    return this.props.status.reduce((healthy: boolean, compStatus: ComponentStatus) => {
      return healthy && compStatus.status === Status.Healthy;
    }, true);
  };

  render() {
    if (!this.healthyComponents()) {
      const icons = this.props.icons ? { ...defaultIcons, ...this.props.icons } : defaultIcons;
      const iconColor = this.tooltipColor();
      let Icon: React.ComponentClass<SVGIconProps> = ResourcesFullIcon;

      if (iconColor === PFColors.Danger) {
        Icon = icons.ErrorIcon;
      } else if (iconColor === PFColors.Warning) {
        Icon = icons.WarningIcon;
      } else if (iconColor === PFColors.Info) {
        Icon = icons.InfoIcon;
      } else if (iconColor === PFColors.Success) {
        Icon = icons.HealthyIcon;
      }

      return (
        <Tooltip position={TooltipPosition.left} enableFlip={true} content={this.tooltipContent()} maxWidth={'25rem'}>
          <Icon color={iconColor} style={{ verticalAlign: '-0.2em', marginRight: -8 }} />
        </Tooltip>
      );
    }

    return null;
  }
}

const mapStateToProps = (state: KialiAppState) => ({
  status: istioStatusSelector(state),
  namespaces: namespaceItemsSelector(state)
});

const mapDispatchToProps = (dispatch: KialiDispatch) => ({
  setIstioStatus: bindActionCreators(IstioStatusActions.setinfo, dispatch),
  refreshNamespaces: () => {
    dispatch(NamespaceThunkActions.fetchNamespacesIfNeeded());
  }
});

export const IstioStatus = connectRefresh(connect(mapStateToProps, mapDispatchToProps)(IstioStatusComponent));
