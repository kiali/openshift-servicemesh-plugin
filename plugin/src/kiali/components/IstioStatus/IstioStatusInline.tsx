import * as React from 'react';
import { IstioStatus } from './IstioStatus';
import {
  CheckCircleIcon,
  ExclamationCircleIcon,
  ExclamationTriangleIcon,
  MinusCircleIcon
} from '@patternfly/react-icons';

type Props = {
  cluster?: string;
};

export function IstioStatusInline({ cluster }: Props) {
  return (
    <IstioStatus
      icons={{
        ErrorIcon: ExclamationCircleIcon,
        HealthyIcon: CheckCircleIcon,
        InfoIcon: MinusCircleIcon,
        WarningIcon: ExclamationTriangleIcon
      }}
      cluster={cluster}
    />
  );
}
