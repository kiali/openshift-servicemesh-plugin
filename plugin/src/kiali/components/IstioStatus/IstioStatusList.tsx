import * as React from 'react';
import { List, Text, TextContent, TextVariants } from '@patternfly/react-core';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { IstioComponentStatus } from './IstioComponentStatus';
import { PFColors } from '../Pf/PfColors';
import { kialiStyle } from 'styles/StyleUtils';
import { useKialiTranslation } from 'utils/I18nUtils';

type Props = {
  status: ComponentStatus[];
};

const listStyle = kialiStyle({
  paddingLeft: 0,
  marginTop: 0,
  marginLeft: 0
});

export const IstioStatusList: React.FC<Props> = (props: Props) => {
  const { t } = useKialiTranslation();

  const nonhealthyComponents = (): ComponentStatus[] => {
    return props.status.filter((c: ComponentStatus) => c.status !== Status.Healthy);
  };

  const coreComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => s.is_core);
  };

  const addonComponentsStatus = (): ComponentStatus[] => {
    return nonhealthyComponents().filter((s: ComponentStatus) => !s.is_core);
  };

  const renderComponentList = (): React.ReactNode => {
    const groups = {
      core: coreComponentsStatus,
      addon: addonComponentsStatus
    };

    return ['core', 'addon'].map((group: string) => {
      return (
        <React.Fragment key={`status-${group}`}>
          {groups[group]().map((status: ComponentStatus) => {
            return <IstioComponentStatus key={`status-${group}-${status.name}`} componentStatus={status} />;
          })}
        </React.Fragment>
      );
    });
  };

  return (
    <TextContent style={{ color: PFColors.White }}>
      <Text component={TextVariants.h4}>{t('Istio Components Status')}</Text>

      <List id="istio-status" aria-label={t('Istio Component List')} className={listStyle}>
        {renderComponentList()}
      </List>
    </TextContent>
  );
};
