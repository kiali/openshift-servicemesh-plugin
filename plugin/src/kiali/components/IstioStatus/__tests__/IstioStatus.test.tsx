import * as React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { ComponentStatus, Status } from '../../../types/IstioStatus';
import { IstioStatusComponent } from '../IstioStatus';
import { shallowToJson } from 'enzyme-to-json';

const mockIcon = (componentList: ComponentStatus[]): ShallowWrapper => {
  return shallow(
    <IstioStatusComponent
      status={componentList}
      lastRefreshAt={848152}
      namespaces={[{ name: 'bookinfo' }, { name: 'istio-system' }]}
      setIstioStatus={jest.fn()}
      refreshNamespaces={jest.fn()}
    />
  );
};

const testSnapshot = (wrapper: any): void => {
  expect(shallowToJson(wrapper)).toBeDefined();
  expect(shallowToJson(wrapper)).toMatchSnapshot();
};

const testTooltip = (wrapper: any): void => {
  expect(wrapper.name()).toEqual('Tooltip');
  expect(wrapper.props().position).toEqual('left');
  expect(wrapper.props().enableFlip).toEqual(true);
  expect(wrapper.children().length).toEqual(1);
};

const testIcon = (wrapper: any, dataTest: string): void => {
  const icon = wrapper.childAt(0);
  expect(icon).toBeDefined();
  expect(icon.props()['data-test']).toEqual(dataTest);
  expect(icon.childAt(0).name()).toEqual('ResourcesFullIcon');
};

describe('When core component has a problem', () => {
  it('the Icon shows is displayed in Red', () => {
    const wrapper = mockIcon([
      {
        name: 'grafana',
        status: Status.Healthy,
        is_core: false
      },
      {
        name: 'istio-egressgateway',
        status: Status.Unhealthy,
        is_core: true
      }
    ]);

    testSnapshot(wrapper);
    testTooltip(wrapper);
    testIcon(wrapper, 'istio-status-danger');
  });
});

describe('When addon component has a problem', () => {
  it('the Icon shows is displayed in orange', () => {
    const wrapper = mockIcon([
      {
        name: 'grafana',
        status: Status.Unhealthy,
        is_core: false
      },
      {
        name: 'istio-egressgateway',
        status: Status.Healthy,
        is_core: true
      }
    ]);

    testSnapshot(wrapper);
    testTooltip(wrapper);
    testIcon(wrapper, 'istio-status-warning');
  });
});

describe('When both core and addon component have problems', () => {
  describe('any component is in not ready', () => {
    it('the Icon shows is displayed in red', () => {
      const wrapper = mockIcon([
        {
          name: 'grafana',
          status: Status.Unhealthy,
          is_core: false
        },
        {
          name: 'istio-egressgateway',
          status: Status.Unhealthy,
          is_core: true
        }
      ]);

      testSnapshot(wrapper);
      testTooltip(wrapper);
      testIcon(wrapper, 'istio-status-danger');
    });
  });
});

describe('When there are not-ready components', () => {
  describe('mixed with other not healthy components', () => {
    describe('in core', () => {
      it('the Icon is displayed in red', () => {
        const wrapper = mockIcon([
          {
            name: 'istio-egressgateway',
            status: Status.Unhealthy,
            is_core: true
          },
          {
            name: 'istio-ingressgateway',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-danger');
      });
    });

    describe('in addons', () => {
      it('the Icon is displayed in orange', () => {
        const wrapper = mockIcon([
          {
            name: 'grafana',
            status: Status.Unhealthy,
            is_core: false
          },
          {
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-warning');
      });
    });

    describe('in both', () => {
      it('the Icon shows is displayed in red', () => {
        const wrapper = mockIcon([
          {
            name: 'grafana',
            status: Status.Unhealthy,
            is_core: false
          },
          {
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          },
          {
            name: 'istio-egressgateway',
            status: Status.Unhealthy,
            is_core: true
          },
          {
            name: 'istio-ingressgateway',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-danger');
      });
    });
  });

  describe('not mixed with other unhealthy components', () => {
    describe('in core', () => {
      it('renders the Icon in blue', () => {
        const wrapper = mockIcon([
          {
            name: 'jaeger',
            status: Status.NotReady,
            is_core: false
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-info');
      });
    });

    describe('in addons', () => {
      it('renders the Icon in blue', () => {
        const wrapper = mockIcon([
          {
            name: 'istiod',
            status: Status.NotReady,
            is_core: true
          }
        ]);

        testSnapshot(wrapper);
        testTooltip(wrapper);
        testIcon(wrapper, 'istio-status-info');
      });
    });
  });
});

describe('When all components are good', () => {
  it('the Icon shows is displayed in green', () => {
    const wrapper = mockIcon([
      {
        name: 'grafana',
        status: Status.Healthy,
        is_core: false
      },
      {
        name: 'istio-egressgateway',
        status: Status.Healthy,
        is_core: true
      }
    ]);

    expect(wrapper.isEmptyRender()).toBeTruthy();
  });
});
