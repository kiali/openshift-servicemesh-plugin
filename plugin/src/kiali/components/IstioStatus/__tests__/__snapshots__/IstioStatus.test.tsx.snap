// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When addon component has a problem the Icon shows is displayed in orange 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "grafana",
            "status": "Unhealthy",
          },
          Object {
            "is_core": true,
            "name": "istio-egressgateway",
            "status": "Healthy",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--warning-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When both core and addon component have problems any component is in not ready the Icon shows is displayed in red 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "grafana",
            "status": "Unhealthy",
          },
          Object {
            "is_core": true,
            "name": "istio-egressgateway",
            "status": "Unhealthy",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--danger-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When core component has a problem the Icon shows is displayed in Red 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "grafana",
            "status": "Healthy",
          },
          Object {
            "is_core": true,
            "name": "istio-egressgateway",
            "status": "Unhealthy",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--danger-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in addons the Icon is displayed in orange 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "grafana",
            "status": "Unhealthy",
          },
          Object {
            "is_core": false,
            "name": "jaeger",
            "status": "NotReady",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--warning-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in both the Icon shows is displayed in red 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "grafana",
            "status": "Unhealthy",
          },
          Object {
            "is_core": false,
            "name": "jaeger",
            "status": "NotReady",
          },
          Object {
            "is_core": true,
            "name": "istio-egressgateway",
            "status": "Unhealthy",
          },
          Object {
            "is_core": true,
            "name": "istio-ingressgateway",
            "status": "NotReady",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--danger-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When there are not-ready components mixed with other not healthy components in core the Icon is displayed in red 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": true,
            "name": "istio-egressgateway",
            "status": "Unhealthy",
          },
          Object {
            "is_core": true,
            "name": "istio-ingressgateway",
            "status": "NotReady",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--danger-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When there are not-ready components not mixed with other unhealthy components in addons renders the Icon in blue 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": true,
            "name": "istiod",
            "status": "NotReady",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--info-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;

exports[`When there are not-ready components not mixed with other unhealthy components in core renders the Icon in blue 1`] = `
<Tooltip
  content={
    <IstioStatusList
      status={
        Array [
          Object {
            "is_core": false,
            "name": "jaeger",
            "status": "NotReady",
          },
        ]
      }
    />
  }
  enableFlip={true}
  maxWidth="25rem"
  position="left"
>
  <ResourcesFullIcon
    color="var(--pf-global--info-color--100)"
    noVerticalAlign={false}
    size="sm"
    style={
      Object {
        "marginRight": -8,
        "verticalAlign": "-0.2em",
      }
    }
  />
</Tooltip>
`;
